components/popover/popover-mixin.js

```diff
@@ -127,68 +215,375 @@ export const PopoverMixin = superclass => class extends superclass {
+		while (node) {
+
+			// observe scrollables
+			let observeScrollable = false;
+			if (node.nodeType === Node.ELEMENT_NODE) {
+				observeScrollable = isScrollable(node, 'overflow-y') || isScrollable(node, 'overflow-x');
+			} else if (node.nodeType === Node.DOCUMENT_NODE) {
+				observeScrollable = true;
+			}
+			if (observeScrollable) {
+				this._scrollablesObserved.push(node);
+				node.addEventListener('scroll', this.#repositionBound);
+			}
+
+			// observe mutations on each DOM scope (excludes sibling scopes... can only do so much)
+			if (node.nodeType === Node.DOCUMENT_NODE || (node.nodeType === Node.DOCUMENT_FRAGMENT_NODE && node.host)) {
+				this._ancestorMutationObserver.observe(node, mutationConfig);
+			}
+
+			node = getComposedParent(node);
+		}
+
+		this._openerIntersectionObserver = new IntersectionObserver(entries => {
+			entries.forEach(entry => this._offscreen = !entry.isIntersecting);
+		}, { threshold: 0 }); // 0-1 (0 -> intersection requires any pixel visible, 1 -> intersection requires all pixels visible)
+		if (this._opener) {
+			this._openerIntersectionObserver.observe(this._opener);
+		}
+
+	}
+
+	#clearDismissible() {
 		if (!this._dismissibleId) return;
 		clearDismissible(this._dismissibleId);
 		this._dismissibleId = null;
 	}
 
-	_focusContent(container) {
+	#constrainSpaceAround(spaceAround, spaceRequired, openerRect) {
+		const constrained = { ...spaceAround };
+
+		const isRTL = this.getAttribute('dir') === 'rtl';
+		if ((this._preferredPosition.span === 'end' && !isRTL) || (this._preferredPosition.span === 'start' && isRTL)) {
+			constrained.left = Math.max(0, spaceRequired.width - (openerRect.width + spaceAround.right));
+		} else if ((this._preferredPosition.span === 'end' && isRTL) || (this._preferredPosition.span === 'start' && !isRTL)) {
+			constrained.right = Math.max(0, spaceRequired.width - (openerRect.width + spaceAround.left));
+		}
+
+		return constrained;
+	}
+
+	#focusContent(container) {
 		if (this._noAutoFocus || this._applyFocus === false) return;
 
 		const focusable = getFirstFocusableDescendant(container);
 		if (focusable) {
-			// Removing the rAF call can allow infinite focus looping to happen in content using a focus trap
+			// removing the rAF call can allow infinite focus looping to happen in content using a focus trap
 			requestAnimationFrame(() => focusable.focus());
 		} else {
-			const content = this._getContentContainer();
+			const content = this.#getContentContainer();
 			content.setAttribute('tabindex', '-1');
 			content.focus();
 		}
 	}
 
-	_focusOpener() {
+	#focusOpener() {
 		if (!document.activeElement) return;
 		if (!isComposedAncestor(this, getComposedActiveElement())) return;
 
 		this?._opener.focus();
 	}
 
-	_getContentContainer() {
-		return this.shadowRoot.querySelector('.content');
+	#getContentContainer() {
+		return this.shadowRoot.querySelector('.content-container');
+	}
+
+	#getLocation(spaceAround, spaceAroundScroll, spaceRequired) {
```

Comment:
This method basically defines where we will place the popover based on the specified preferred location, based on the space available in the viewport and document.